lib/checkleakautovar.cpp

```diff
@@ -553,6 +553,53 @@ void CheckLeakAutoVar::checkScope(const Token * const startToken,
         else if (Token::Match(tok, "continue|break ;")) {
             varInfo->clear();
         }
+
+        // Check smart pointer
+        else if (Token::Match(ftok, "auto_ptr|unique_ptr|shared_ptr < %type%")) {
+
+
+            const Token * endToken = ftok->linkAt(1);
+
+            bool arrayDelete = false;
+            if(Token::findsimplematch(ftok->tokAt(1), "[ ]", endToken))
```

Comment:
I don't like the order here. Do the `Token::Match` on endToken immediately. If it is null you are potentially scanning the entire source code with findsimplematch.